### PR TITLE
Improve Rebel Weapon and Other Items Selection

### PR DESCRIPTION
**Pull request by jwoodruff40**
## What type of PR is this?
1. [ ] Bug
2. [X] Change
3. [X] Enhancement
4. [ ] Miscellaneous

### What have you changed and why?
Information: 

**Why:**

It was bothering me that when you start to accumulate / unlock better weapons in your arsenal, AI Rebels will just randomly 'pick' any weapon in a given category, instead of trying to 'pick' a better weapon. E.g. when playing as 3CB FIA, my rebels will quite often be equipped with M1 Garands (initial equipment) even when I have M16s, AK74s, AK105s, etc unlocked. This severely limits the Rebels' firepower compared to enemy AI.

I also noticed that fn_generateRebelGear / the A3A_rebelGear it creates is structured as a weighted list but doesn't currently make use of weights, adding everything with a weight of 1.

**What:**

This update to fn_generateRebelGear calculates a weaponWeight for each primary weapon added to A3A_rebelGear based on the weapon's attributes (positive and negative):

Positive attributes (higher is better):

 - accuracy
 - rate of fire
 - magazine capacity
 - 'impact force'
 - effective range (currently only for snipers and launchers)
 - penetration (launchers)
 
Negative attributes (higher is worse):

 - weight
 - disposable (launchers)

It's heavily inspired by the weapon statistics shown in ACE3 Arsenal (especially the impact attribute), but the only code copied from ACE is the impact force calculation, a combination of muzzle velocity and bullet / ammo damage.

The end state is that when AI rebels are equipped, they'll be much more likely to equip better weapons, without entirely precluding use of 'lower tier' weapons. To that end, I attempted to weigh the weapon attributes sensibly.

Example weapon weights from a pretty new save (not very many weapons) with 3CB FIA vs 3CB AAF (some output omitted):

`A3A_rebelGear;` ==> 

```
[
[["GrenadeLaunchers",[
    "UK3CB_M16A2_UGL",40
]],
["MissileLaunchersAA",[
    "rhs_weap_fim92",
    "UK3CB_Blowpipe"
]],
["Shotguns",[
    "rhs_weap_M590_5RD",7
]],
["SniperRifles",[
    "UK3CB_Enfield",10,
    "UK3CB_M1903A1",3
]],
["Rifles,[
    "rhs_weap_m1garand_sa43",16,
    "rhs_weap_ak74m",45,
    "UK3CB_M16A2",47,
    "UK3CB_HK33KA1",40,
    "UK3CB_HK33KA3",40,
    "rhs_weap_ak105",53
]],
["SMGs",["rhs_weap_m3a1",53]]
]
```

### Please specify which Issue this PR Resolves (If Applicable).
None as far as I'm aware.

### Please verify the following.

1. [X] Have you loaded the mission in LAN host?
2. [ ] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?

1. [ ] No
2. [X] Yes (Please provide further detail below.)

~~Modification has only been tested with RHS and 3CBF factions / weapons, with a few randomly picked selections of test weapons and on a couple save files that use those factions.~~ Needs more testing to ensure compatibility across other supported factions / weapon mods. I would also be interested in expanding to include weighing launchers and potentially other items, but need feedback on if that's wanted and what stats would make sense to weigh those items. 

### How can the changes be tested?
Steps:

Best way to test is probably to get a bunch of people running different factions and such to copy their save(s), load the copied save, and in the debug console input `A3A_rebelGear;`  and analyze the output for discrepancies, weapons whose stats / weight weren't properly identified / calculated, or if the weights seem to be either too close or too disparate strongly favoring certain weapons or not.

********************************************************
Notes:

~~I wanted to include weapon max range (or max effective range?) as an additional attribute, but I found that at least for RHS weapons the most ideal attributes to use (_config >> maxRange / _config >> maxRangeProbab) are generally all the same in the weapon config root, and are only updated under the fire modes, but a lot of RHS weapons use several custom fire modes instead of the normal "SINGLE" or "BURST" or "FULLAUTO", and I didn't want to have to iterate through all fire modes to attempt to find the best values like is done in ACE. Ultimately I decided that the impact force alone should be sufficient, given that generally, longer range weapons have higher muzzle velocity / larger bullets - the same things that affect impact force. May be worth revisiting though.~~

Effective range is now calculated, but loops through all firemodes so probably best not to use for every weapon type.

### CODE REVIEW

**Comment by jwoodruff40** [comment]
Another thought I had – I don't play with unlocks disabled, but for those that do, the amount of each weapon in the arsenal could also be included in the weapon weight calculation such that AI will be more likely to be equipped with weapons the rebels have more of, all else being equal.

**Comment by wersal454** [comment]
That looks really nice

For those who play with unlocks disabled, system I think should be adapted yes(if you can figure it out ofc)

Also maybe add a weights for secondary weapons(i.e. launchers)? Not sure how to qualify some launchers to be better then others, but maybe take a look at it
also I want to add handguns in #418, so maybe weights system for them as well?

**Comment by Silence-112** [comment]
The base branch should be unstable rather than stable, changing it may cause merge conflicts so be aware of those

**Comment by wersal454** [comment]
just in case

Click "Edit" and this menu will pop up
image
there you can select different base branch

**Comment by jwoodruff40** [comment]
Thanks team for the input. I changed the base branch.

I have some thoughts on how to go about weighing other item types, but don't want to get too into the weeds lest it starts becoming too much of a subjective / playstyle-dependent approach. More to follow.

Merry Christmas / happy holidays to all and thanks for the awesome mod!

**Comment by jwoodruff40** [comment]
I fixed my rookie mistake of changing the param order for _fnc_addItem  and did some initial rework of the existing code for less code reuse.

Also pushed an initial implementation for _fnc_addItemNoUnlocks. It uses the already existing linearconversion of amount of item to range between 0 and 1 as a scalar for the item weight variable. So items with similar weight will be selected more often if there are more of them, and vice versa. Should make for some interesting results with varying item weights and amounts. 

Personally, I think that's a good enough implementation, but I don't ever play with unlocks disabled, so we could use some input from the crowd that do.

Edit: I am curious though - why do we divide the amount of rifles (and only rifles) by 2 when passing the amount to the addItem function? This does significantly skew how many you need to get for AI to realistically start using them, which I guess is maybe the intent? But then why not for other types of primary weapons?

**Comment by jwoodruff40** [comment]
Refactored / generalized function to support whatever item types we want to implement custom weight / preference calculation for (secondaries, handguns, armor, etc).

**Comment by UAC-MaxxLite** [comment]
I would also suggest that you limit the type of secondaries for each class if possible I.e. 
only AT and AA units being able to use guided launchers.
And disable secondary use for marksman medic and such.

This will make it harder for players that spam launchers on all types of units which just makes the game easier.

**Comment by jwoodruff40** [comment]
@UAC-MaxxLite that's done already in fn_equipRebel unless I'm misunderstanding what you mean. If you mean restricting players from override the AI loadouts with that kind of thing, I get what you mean. Not sure if that's in the scope of this PR or not though.

In any case, latest commit implements specific weight calculations for sniper rifles, rocket launchers (LAT), and handguns (per @wersal454 's suggestion / PR). How and why the attributes are weighted the way they are for those is commented in the file, but as stated before, definitely needs a lot of testing across supported mod sets.

The func is getting pretty long and will only get longer with more variable / calculations for additional item types, so I also broke it out into its own file.

**Comment by wersal454** [comment]
> I would also suggest that you limit the type of secondaries for each class if possible I.e. only AT and AA units being able to use guided launchers. And disable secondary use for marksman medic and such.
> 
> This will make it harder for players that spam launchers on all types of units which just makes the game easier.

restricting players from equipping any AI with launchers would take A TON of work, it would require changing arsenal code(probably) which is a bitch

**Review by Silence-112 — COMMENTED** [review]
Looks fine, can be tested when you feel it ready

**Comment by jwoodruff40** [comment]
Pushed a few more fixes and things. I would call it ready for testing now. Can't (or don't know how) to set that label myself.

In summary (things to look for with testing the code), this PR does two main things:
 - Makes rebel forces (personal and HC squads) more likely to equip better weapons in each category, while still maintaining a equipping a variety of weaponry
 - Equips some rebel rifleman (percentage increases by war level) with disposable AT launcher

And sets the conditions for crafting interesting weights for other item types in the future - e.g. preferencing vests by some combination of weight, load capacity, armor value. 

@wersal454 – I did some minor manual testing of weighting for handguns just to make sure they looked kinda right, but if you want to give this a test along with your PR that actually equips rebels with handguns, that would be helpful.

---

I also pushed a larger refactor of `fn_generateRebelGear` (last commit). Intent was to remove a lot of the duplicate code between item types. Understand code style is very subjective and how it is now is probably more readable (if much more verbose), so can revert if not wanted. Can test with or without that commit.

**Review by Silence-112 — COMMENTED** [review]
Code looks fine

We'll try to test this in our next server campaign, will help bring any potential issues to light

**Review by Silence-112 — CHANGES_REQUESTED** [review]
Testing this locally reveals that no AI will actually get weapons due to errors, even Petros doesn't get given a weapon or vest. 

I would assume this is due to the fact that the _rebelGear hashmap never has any data set to it anymore, including no ArmoredVests/ArmoredHeadgear

```sqf
15:18:53 Error in expression <A_rebelGear get "ArmoredHeadgear");
if (_helmet == "") then { _helmet = selectRa>
15:18:53   Error position: <_helmet == "") then { _helmet = selectRa>
15:18:53   Error Undefined variable in expression: _helmet
15:18:53 File x\A3A\addons\core\functions\REINF\fn_equipRebel.sqf..., line 96
15:18:53  ➥ Context: 	[] L35 (x\A3A\addons\core\functions\REINF\fn_FIAinit.sqf)
	[] L36 (x\A3A\addons\core\functions\REINF\fn_FIAinit.sqf)
	[] L108 (x\A3A\addons\core\functions\REINF\fn_equipRebel.sqf)

15:18:53 Error in expression <
private _compatItems = compatibleItems _weapon; 
_compatOptics = _compatItems a>
15:18:53   Error position: <_weapon; 
_compatOptics = _compatItems a>
15:18:53   Error Undefined variable in expression: _weapon
15:18:53 File x\A3A\addons\core\functions\Ammunition\fn_randomRifle.sqf..., line 66
15:18:53  ➥ Context: 	[] L35 (x\A3A\addons\core\functions\REINF\fn_FIAinit.sqf)
	[] L36 (x\A3A\addons\core\functions\REINF\fn_FIAinit.sqf)
	[] L111 (x\A3A\addons\core\functions\REINF\fn_equipRebel.sqf)
	[] L116 (x\A3A\addons\core\functions\REINF\fn_equipRebel.sqf)
	[] L65 (x\A3A\addons\core\functions\Ammunition\fn_randomRifle.sqf)
	[] L67 (x\A3A\addons\core\functions\Ammunition\fn_randomRifle.sqf)
	[] L68 (x\A3A\addons\core\functions\Ammunition\fn_randomRifle.sqf)

15:18:53 Error in expression <ol = A3A_rebelGear get _weaponType;
if (_pool isEqualTo []) then {
_pool = A3A_r>
15:18:53   Error position: <_pool isEqualTo []) then {
_pool = A3A_r>
15:18:53   Error Undefined variable in expression: _pool
15:18:53 File x\A3A\addons\core\functions\Ammunition\fn_randomRifle.sqf..., line 24
15:18:53  ➥ Context: 	[] L35 (x\A3A\addons\core\functions\REINF\fn_FIAinit.sqf)
	[] L36 (x\A3A\addons\core\functions\REINF\fn_FIAinit.sqf)
	[] L111 (x\A3A\addons\core\functions\REINF\fn_equipRebel.sqf)
	[] L116 (x\A3A\addons\core\functions\REINF\fn_equipRebel.sqf)
	[] L65 (x\A3A\addons\core\functions\Ammunition\fn_randomRifle.sqf)
	[] L67 (x\A3A\addons\core\functions\Ammunition\fn_randomRifle.sqf)
	[] L68 (x\A3A\addons\core\functions\Ammunition\fn_randomRifle.sqf)

15:18:53 Error in expression <A3A_rebelGear get "OpticsClose";
};
if (_compatOptics isEqualTo []) then {
_comp>
15:18:53   Error position: <_compatOptics isEqualTo []) then {
_comp>
15:18:53   Error Undefined variable in expression: _compatoptics
15:18:53 File x\A3A\addons\core\functions\Ammunition\fn_randomRifle.sqf..., line 72
15:18:53  ➥ Context: 	[] L35 (x\A3A\addons\core\functions\REINF\fn_FIAinit.sqf)
	[] L36 (x\A3A\addons\core\functions\REINF\fn_FIAinit.sqf)
	[] L111 (x\A3A\addons\core\functions\REINF\fn_equipRebel.sqf)
	[] L116 (x\A3A\addons\core\functions\REINF\fn_equipRebel.sqf)
	[] L65 (x\A3A\addons\core\functions\Ammunition\fn_randomRifle.sqf)
	[] L78 (x\A3A\addons\core\functions\Ammunition\fn_randomRifle.sqf)

15:18:53 Error in expression <he set [_weapon, _compatOptics];
};
if (_compatOptics isNotEqualTo []) then { _u>
15:18:53   Error position: <_compatOptics isNotEqualTo []) then { _u>
15:18:53   Error Undefined variable in expression: _compatoptics
15:18:53 File x\A3A\addons\core\functions\Ammunition\fn_randomRifle.sqf..., line 80
15:18:53  ➥ Context: 	[] L35 (x\A3A\addons\core\functions\REINF\fn_FIAinit.sqf)
	[] L36 (x\A3A\addons\core\functions\REINF\fn_FIAinit.sqf)
	[] L111 (x\A3A\addons\core\functions\REINF\fn_equipRebel.sqf)
	[] L116 (x\A3A\addons\core\functions\REINF\fn_equipRebel.sqf)
	[] L80 (x\A3A\addons\core\functions\Ammunition\fn_randomRifle.sqf)

15:18:53 Error in expression <eighted (A3A_rebelGear get "NVGs");
if (_nvg != "") then { _unit linkItem _nvg }>
15:18:53   Error position: <_nvg != "") then { _unit linkItem _nvg }>
15:18:53   Error Undefined variable in expression: _nvg
15:18:53 File x\A3A\addons\core\functions\REINF\fn_equipRebel.sqf..., line 222
15:18:53  ➥ Context: 	[] L35 (x\A3A\addons\core\functions\REINF\fn_FIAinit.sqf)
	[] L36 (x\A3A\addons\core\functions\REINF\fn_FIAinit.sqf)
	[] L241 (x\A3A\addons\core\functions\REINF\fn_equipRebel.sqf)
```

**Comment by jwoodruff40** [comment]
> Testing this locally reveals that no AI will actually get weapons due to errors, even Petros doesn't get given a weapon or vest.
> 
> I would assume this is due to the fact that the _rebelGear hashmap never has any data set to it anymore, including no ArmoredVests/ArmoredHeadgear

~~Not to be the developer that says "it works for me, must be a you problem" at all, but.. I'm not getting any errors and it seems to work fine with no mods loaded except A3U built with this PR, CBA, and ace. But if we can't figure it out, we can just revert https://github.com/SilenceIsFatto/A3-Antistasi-Ultimate/pull/427/commits/46c81fa3923aa0dd4078ea4c10800b6debe378f0.~~

**\*\*\*EDIT:\*\*\*** this is what I get for thinking to test with minimal mods, but *not* testing a **brand new game.** >facepalm<

This is an issue where the refactor only adds the categories when you have items in them in the first place (initial rebel gear from rebel template). So I / we can either -

1) initialize the hashmap with empty arrays for *every* category, e.g. `private _rebelGear = createHashMap;` ==> 

```
private _rebelGear = createHashMapFromArray [
    ["Rifles", []],
    ["SMGs", []],
    ["Shotguns", []],
    ["MachineGuns", []],
    ["SniperRifles", []],
    ["GrenadeLaunchers", []],
    
    ["RocketLaunchers", []],
    ["MissileLaunchersAT", []],
    ["MissileLaunchersAA", []],

    ["ArmoredVests", ["", [1.5, 0.5] select (minWeaps < 0)]],
    ["CivilianVests", []],
    ["ArmoredHeadgear", ["", [1.5, 0.5] select (minWeaps < 0)]],
    ["BackpacksCargo", []],

    ["NVGs", ["", 0.5]],
    ["Radios", []],
    ["MineDetectors", []],
    ["Toolkits", []],
    ["Medikits", []],

    ["SmokeGrenades", []],
    ["Grenades", []],
    ["ExplosiveCharges", []],

    ["OpticsClose", []],
    ["OpticsMid", []],
    ["OpticsLong", []],
    ["OpticsAll", []],
    ["LightAttachments", []]
]; 
```

and the rest of the code *should* work as normal. ~~I'll submit a suggested change for ^^ after I finish a bit of my day job work and make sure it works, to~~ @SilenceIsFatto -  https://github.com/SilenceIsFatto/A3-Antistasi-Ultimate/pull/427/commits/cc2c372815d2479d5665f2f33108ee9e0b49a65f fixes the issue on my end. Thoughts on this vs option 2)?.

**OR**

2) revert https://github.com/SilenceIsFatto/A3-Antistasi-Ultimate/pull/427/commits/46c81fa3923aa0dd4078ea4c10800b6debe378f0 for the time being.

Leaving the below explanation in case it's helpful to someone for thought process?

**\*\*\*END EDIT\*\*\***

---

The gist of the refactor is that all instances of a code block like this:

```
private _itemType = []; // internal array
{
    _item = _x;
    <do some stuff>
    if (condition) then { _itemType append [_item, _itemWeight] } ; // add the item to the internal array
} forEach _items;

_rebelGear set ["Item Type", _itemType]; // set the hashmap from the pre-populated internal array
```

are instead replaced with:

```
// instead of using an internal array and pushing it to the hashmap when we're done, push directly to the hashmap
// if the item type doesn't yet exist in the hashmap, add an empty array
private _itemType = _rebelGear getOrDefault ["Item Type", [], true];
{
    _item = _x;
    <do some stuff>
    if (condition) then { _itemType append [_item, _itemWeight] } ; // add the item directly to the hashmap array
} forEach _items;
```

In short - `_rebelGear set [key, val]` is replaced with `_rebelGear getOrDefault [key, [], true]`.  `setDefault` param `true` creates the hashmap entry. See [getOrDefault](https://community.bistudio.com/wiki/getOrDefault)

**Review by Silence-112 — COMMENTED** [review]
Works as expected now, seems to be much better than the previous system.

It'd be nice if A3A_rebelGear could be re-evaluated as you unlocked things rather than once you load a save, but I imagine there is a reason this is done once during init.

Regardless, looks fine now. I'll do a bit more testing with the obscure modsets like OPTRE and if it works then this should be fine for merge

**Comment by jwoodruff40** [comment]
> It'd be nice if A3A_rebelGear could be re-evaluated as you unlocked things rather than once you load a save, but I imagine there is a reason this is done once during init.

fn_generateRebelGear updates A3A_rebelGear and is called by fn_resourceCheck once on server init and every 10 minutes thereafter (fn_initServer line 276, fn_resourceCheck line 185).

Afaict, there's no issues with calling it whenever you want as server admin.

It should be easily possible to *also* call generateRebelGear anytime an item is either unlocked (unlocks enabled) or quantity surpasses ITEM_MIN (no unlocks). But in the meantime, if I'm not mistaken checking for victory manually at the map board / in the commander menu calls resourceCheck, so that would be an in-game way to force update rebel gear without having to call generateRebelGear from the debug console.

**Review by Silence-112 — APPROVED** [review]
Was thinking of when things transfer to the arsenal but with the resourceCheck loop that would be redundant

It's much better with vanilla, seems to be better with OPTRE and works fine with spearhead 1944. I think it's fine for merge
